### PR TITLE
[WIP] Protobuf Guide

### PR DESCRIPTION
**Pull request by breerly**
This part of a Red Team exercise with @peter-edge, and my first time using Protobuf in YARPC 👍 

This PR represents the Protobuf Guide for #1241 

- [x] Sketch out initial server, capturing notes along the way
- [x] Sketch out initial client, capturing notes along the way
- [ ] Refine client and server code
- [ ] Finish the Proto Guide

### CODE REVIEW

File: .docs/08-binary-encodings.md [hunk 1]
```diff
@@ -4,6 +4,188 @@
 
 Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur eget commodo magna. Nam varius, purus tincidunt pulvinar auctor, sapien lorem consequat tortor, in cursus enim sapien vel erat. Donec non sollicitudin neque. Integer rutrum egestas purus ac porttitor. Morbi maximus metus dapibus iaculis molestie. Nam euismod sed est quis auctor. Aenean at odio vitae orci sodales lobortis nec a nibh. Maecenas condimentum ipsum risus, eu pellentesque augue facilisis a. In hac habitasse platea dictumst. Quisque fermentum accumsan felis, eu venenatis nisl pulvinar vel. Nam eu accumsan est. Sed viverra vel ligula sit amet placerat. Donec molestie ut mauris nec egestas.
```
Comment (peter-edge) [inline .docs/08-binary-encodings.md]: Remove this

File: .docs/08-binary-encodings.md [hunk 2]
```diff
@@ -4,6 +4,188 @@
...
 
+## Thrift
+
+Walk end-user through building the thrift-hello example.
+
+## Protobuf
```
Comment (peter-edge) [inline .docs/08-binary-encodings.md]: Need to add installing protoc, I can help with this

File: .docs/08-binary-encodings.md [hunk 3]
```diff
@@ -4,6 +4,188 @@
...
+
+```
+$ go get github.com/gogo/protobuf/protoc-gen-gogoslick
+$ go get go.uber.org/yarpc/encoding/protobuf/protoc-gen-yarpc-go
+```
+
+Author a protobuf IDL:
+
+```proto
+syntax = "proto3";
```
Comment (peter-edge) [inline .docs/08-binary-encodings.md]: Do style guide

File: .docs/08-binary-encodings.md [hunk 4]
```diff
@@ -4,6 +4,188 @@
...
+Install protoc and YARPC's proto plugin:
+
+```
+$ go get github.com/gogo/protobuf/protoc-gen-gogoslick
+$ go get go.uber.org/yarpc/encoding/protobuf/protoc-gen-yarpc-go
+```
+
+Author a protobuf IDL:
+
+```proto
```
Comment (peter-edge) [inline .docs/08-binary-encodings.md]: Prob want `package hello, service Bar, rpc Foo`

File: .docs/08-binary-encodings.md [hunk 5]
```diff
@@ -4,6 +4,188 @@
...
+<-signals
+```
+
+Start the server:
+
+```
+$ go run ./server/main.go
+```
+
+Curl the server:
```
Comment (peter-edge) [inline .docs/08-binary-encodings.md]: Mention that the protobuf encoding implementation in yarpc-go handles both protobuf and json encodings

File: .docs/08-binary-encodings.md [hunk 6]
```diff
@@ -4,6 +4,188 @@
...
+	"log"
+	"strings"
+	"time"
+
+	hello "go.uber.org/yarpc/internal/examples/protobuf-hello"
+	"go.uber.org/yarpc/transport/http"
+	"go.uber.org/yarpc/x/config"
+)
+
+const yaml = `
```
Comment (peter-edge) [inline .docs/08-binary-encodings.md]: `yamlData`?

File: internal/examples/protobuf-hello/client/main.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+)
+
+const yaml = `
+outbounds:
+  server:
+    http:
+      url: "http://localhost:8080/"
+`
+
+func main() {
```
Comment (peter-edge) [inline internal/examples/protobuf-hello/client/main.go]: I again register my distaste for this style, would prefer:

```go
func main() {
  if err := do(); err != nil {
    log.Fatal(err)
  }
}
```

And then don't use panic functions. I think this style being in the example sets a good example (get it?).

File: .docs/08-binary-encodings.md (hunk 6 above)
Comment (kriskowal) [inline .docs/08-binary-encodings.md]: Perhaps `config`?

**Comment by peter-edge** [comment]
Closing for now, feel free to reopen if we want to continue on this, I think we're doing in a different direction now

**Pull request closed without merging**